### PR TITLE
Optional fields

### PR DESCRIPTION
**Pull request by phadej**
Continuation of #1023

### CODE REVIEW

**Review by friedbrice — COMMENTED** [review]
I was thinking Aeson could retain backwards compatibility if instead of adding an argument to `liftToJSON`, a method `liftToJSONWithOmit` could be added. It could have the default implementation `liftToJSONOmit _ = liftToJSON` to preserve the behavior of people's existing instances. Similar additions to `ToJSON2` and `FromJSON1`, `FromJSON2`.

**Comment by phadej** [comment]
> I was thinking Aeson could retain backwards compatibility

It's not worth it. There are barely any direct users of these https://hackage-search.serokell.io/?q=liftToJSON and arguably the ones who are (e.g. `autocodec`, `insert-ordered-containers`, should adapt).

**Comment by friedbrice** [comment]
@phadej That all makes a lot of sense. Thanks for explaining the situation to me. (Also, I'll start defining `toEncoding`!)

File: src/Data/Aeson/TH.hs [hunk 1]
```diff
@@ -456,40 +454,26 @@ argsToValue letInsert target jc tvMap opts multiCons
...
+            toPair (arg, argTy, fld) =
+              let fieldName = fieldLabel opts fld
+                  toValue = dispatchToJSON target jc conName tvMap argTy
+
+                  omitFn :: Q Exp
+                  omitFn
+                    | omitNothingFields opts = dispatchOmitField jc conName tvMap argTy
+                    | otherwise = [| const False |]
+              in
+              [| \f x arg' -> bool x mempty (f arg') |]
```
Comment (phadej) [inline src/Data/Aeson/TH.hs:471]: TODO: use `if then else`.

File: tests/PropertyTH.hs [hunk 2]
```diff
@@ -130,10 +130,5 @@ templateHaskellTests =
         thOneConstructorToJSONDefault `sameAs` thOneConstructorToEncodingDefault
       , testProperty "OneConstructorTagged" $
         thOneConstructorToJSONTagged `sameAs` thOneConstructorToEncodingTagged
-
-#if !MIN_VERSION_base(4,16,0)
```
Comment (phadej) [inline tests/PropertyTH.hs:134]: Un remove this?

File: src/Data/Aeson/Types/FromJSON.hs [hunk 3]
```diff
@@ -379,6 +383,33 @@ class FromJSON a where
...
+    -- Defining @omittedField = 'Nothing'@ makes object fields of this type required.
+    --
+    -- @
+    --   omittedField :: Maybe Int -- Nothing
+    --   decode "{\"a\":1}" -- Nothing
+    -- @
+    --
+    -- The default implementation is @omittedField = Nothing@.
+    --
+    -- @since x.x.x.x
```
Comment (phadej) [inline src/Data/Aeson/Types/FromJSON.hs:409]: TODO

File: src/Data/Aeson/Types/FromJSON.hs [hunk 4]
```diff
@@ -1336,34 +1398,55 @@ instance ( RecordFromJSON' arity a
...
+
+
+recordParseJSONImpl :: forall s arity a f i
+                     . (Selector s)
+                    => Maybe (f a)
+                    -> (Options -> FromArgs arity a -> Value -> Parser (f a))
+                    -> (ConName :* TypeName :* Options :* FromArgs arity a)
+                    -> Object -> Parser (M1 i s f a)
+recordParseJSONImpl mdef parseVal (cname :* tname :* opts :* fargs) obj =
+  handleMissingKey (M1 <$> mdef) $ do
```
Comment (phadej) [inline src/Data/Aeson/Types/FromJSON.hs:1440]: can we use `explicitParseFieldOmit` here?

File: src/Data/Aeson/Types/FromJSON.hs [hunk 5]
```diff
@@ -840,6 +877,10 @@ parseFieldMaybe = (.:?)
 parseFieldMaybe' :: (FromJSON a) => Object -> Key -> Parser (Maybe a)
 parseFieldMaybe' = (.:!)
 
+-- | Function variant of '.:?='.
```
Comment (phadej) [inline src/Data/Aeson/Types/FromJSON.hs:880]: `@since` annotation, o nthis and `omittedField2` etc.

**Pull request merged**